[REVIEW] Add dtypes information page in python docs

galipremsagar opened the pull request:
Resolves: #5278 

This PR adds dtypes information for cudf. This is how the page looks:
<img width="1293" alt="Screen Shot 2020-08-14 at 7 49 39 PM" src="https://user-images.githubusercontent.com/11664259/90301749-522e8680-de67-11ea-9e1b-6874282dc21c.png">



1. As we add a dedicated docs pages for each dtype we could start adding a new column named under **Documentation** providing links for the same - similar to how pandas does https://pandas.pydata.org/pandas-docs/stable/user_guide/basics.html#basics-dtypes

2. Do we need **Data Type** column for now? Maybe as we have cudf Dtypes implemented we can populate this column.

cc: @brandon-b-miller @kkraus14 @shwina 

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

brandon-b-miller commented:
I think the Data Type column is good to have, but we should explain what that means and why. Maybe add some explanation about why we have a native `CategoricalDtype` and `ListDtype` where we can't use numpy. Later we can shift it away from numpy and to our own dtypes if we need to.

shwina commented on docs/cudf/source/index.rst:
```diff
@@ -6,6 +6,7 @@ Welcome to cuDF's documentation!
    :caption: Contents:
 
    api.rst
+   basics.rst
```
I think this should come after the 10min notebook(s), as that gives a broad introduction that most of our users are probably looking for first.

raydouglass commented on conda/environments/cudf_dev_cuda10.0.yml:
```diff
@@ -50,6 +50,7 @@ dependencies:
   - hypothesis
   - sphinx-markdown-tables
   - sphinx-copybutton
+  - sphinxemoji
```
Need to add this to the `rapids-doc-env` as well: https://github.com/rapidsai/integration/blob/branch-0.15/conda/recipes/rapids-doc-env/meta.yaml

galipremsagar replied:
Added to integration repo : https://github.com/rapidsai/integration/pull/90

kkraus14 commented on docs/cudf/source/basics.rst:
```diff
@@ -0,0 +1,19 @@
+
+The following table lists all of cudf types. For methods requiring dtype arguments, strings can be specified as indicated. See the respective documentation sections for more on each type.
+
+.. csv-table:: Supported dtypes in cudf
+   :file: dtypes.csv
+   :align: center
+   :widths: 30, 5, 30, 40, 50
+   :header-rows: 1
```
Instead of making this CSV driven is there a way to build the table using markdown / rst formatting instead? It would make updating / maintaining it much smoother.

galipremsagar replied:
Apart from [csv table](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html#csv-files), there two different [ways](https://tables-with-sphinx.readthedocs.io/en/latest/index.html)

1. [List table](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/tables.html#list-table-directive) Everything is vertically stacked so it would be tougher to maintain/edit - Probably more error-prone.
2. [Grid table](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#grid-tables) This would be a one-time effort of when creating and making some major restructuring and easier to make minor edits.

kkraus14 replied:
I'm a big +1 to Grid Table

galipremsagar replied:
I'll try out the Grid table for our setting and will go ahead with the grid table if there are no issues with it.

kkraus14 commented:
I think we can remove the `Nullable` column also and just add a sentence that all of our types are nullable.

galipremsagar commented:
Made some changes by moving to grid table and added listdtype, but I doubt if we have to point to python list scalar as, updated the screenshot of how it looks like. @shwina Maybe you have more insight on what is appropriate representation of scalar incase of `ListDtype`

shwina commented:
> @shwina Maybe you have more insight on what is appropriate representation of scalar & string alias incase of ListDtype

I think what we have right now, i.e., `list` and `"list"` respectively is fine. Thanks!

galipremsagar commented:
rerun tests

kkraus14 commented on docs/cudf/source/basics.rst:
```diff
@@ -0,0 +1,40 @@
+| Timedelta              |            | `numpy.timedelta64 <https://numpy.org/doc/stable/reference/arrays.datetime.html>`_  | ``'timedelta64[s]'``, ``'timedelta64[ms]'``,|
+| (duration type)        |            |                                                                                     | ``'timedelta64[us]'``, ``'timedelta64[ns]'``|
++------------------------+------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Categorical            |            | (none)                                                                              | ``'category'``                              |
```
Should we have `CategoricalDtype` in here since we technically have a separate dtype from Pandas?

kkraus14 commented on docs/cudf/source/basics.rst:
```diff
@@ -0,0 +1,40 @@
++------------------------+------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Kind of Data           | Data Type  | Scalar                                                                              | String Aliases                              |
++========================+============+=====================================================================================+=============================================+
+| Integer                |            | `int <https://docs.python.org/3/library/functions.html#int>`_                       | ``'int8'``, ``'int16'``, ``'int32'``,       |
+|                        |            |                                                                                     | ``'int64'``, ``'uint8'``, ``'uint16'``,     |
+|                        |            |                                                                                     | ``'uint32'``, ``'uint64'``                  |
++------------------------+------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Float                  |            | `float <https://docs.python.org/3/library/functions.html#float>`_                   | ``'float32'``, ``'float64'``                |
```
Do we not use numpy scalars for these?

kkraus14 commented on docs/cudf/source/basics.rst:
```diff
@@ -0,0 +1,40 @@
++------------------------+------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Categorical            |            | (none)                                                                              | ``'category'``                              |
++------------------------+------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Boolean                |            | `bool <https://docs.python.org/3/library/functions.html#bool>`_                     | ``'bool'``                                  |
```
Similarly do we not use a numpy scalar for this?

galipremsagar replied:
Updated the content and links

kkraus14 commented on docs/cudf/source/basics.rst:
```diff
@@ -0,0 +1,54 @@
++------------------------+------------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| Boolean                |                  | np.bool_                                                                            | ``'bool'``                                  |
++------------------------+------------------+-------------------------------------------------------------------------------------+---------------------------------------------+
+| List                   |  ListDtype       | `list <https://docs.python.org/3/tutorial/datastructures.html#more-on-lists>`_      | ``'list'``                                  |
```
Should the string alias here be `List(type)` since we expected a capital `L` as well as expected the child type supplied as well?

galipremsagar replied:
Yes, after looking at the code I should have added `ListDtype("int"), .., ..etc` but  Looks like `ListDtype` is missing docs and not exposed in sphinx-docs, should I add expose it for `0.15`? Is there some limitation of note/disclaimer to be made for the support of the `ListDtype`?

kkraus14 replied:
Lets not expose it for `0.15` as the support is extremely early / not really supported well yet.

galipremsagar replied:
In that case, should I remove the entry from dtypes table?

kkraus14 replied:
I would propose we do that for 0.15 and we can add it in 0.16.

galipremsagar commented:
rerun tests
